fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/ngram/NgramAndSynonymAnalyzerFactory.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * NgramAnalyzer.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene.ngram;
+
+import com.apple.foundationdb.record.lucene.LuceneAnalyzerFactory;
+import com.apple.foundationdb.record.lucene.synonym.SynonymAnalyzer;
+import com.apple.foundationdb.record.metadata.Index;
+import com.google.auto.service.AutoService;
+import org.apache.lucene.analysis.Analyzer;
+
+import javax.annotation.Nonnull;
+import java.util.Objects;
+
+@AutoService(LuceneAnalyzerFactory.class)
+public class NgramAndSynonymAnalyzerFactory implements LuceneAnalyzerFactory {
+    public static final String ANALYZER_NAME = "NGRAM_AND_SYNONYM";
+
+    @Nonnull
+    @Override
+    public String getName() {
+        return ANALYZER_NAME;
+    }
+
+    @SuppressWarnings("deprecation")
+    @Nonnull
+    @Override
+    public Analyzer getIndexAnalyzer(@Nonnull Index index) {
+        return new NgramAnalyzer.NgramAnalyzerFactory().getIndexAnalyzer(index);
+    }
+
+    @Nonnull
+    @Override
+    public Analyzer getQueryAnalyzer(@Nonnull Index index, @Nonnull Analyzer indexAnalyzer) {
```

Comment:
I am afraid the main concern with using ngram analyzer for indexing and synonym analyzer for query is, the synonyms of substrings of indexed terms could be considered as matches to return, which is unexpected. This is just the main problem we need to resolve, to support analyzer for combination of ngram and synonym, right?